### PR TITLE
Support gsd 3.

### PR DESCRIPTION
**Pull request by joaander**
## Description

* Update unit tests to support GSD 3.
* Test with gsd 2.9 and ensure that there are no deprecation warnings.

## Motivation and context

Tests will fail with GSD 3 as these deprecated modes will be removed.

## How has this been tested?

Unit tests pass locally with no gsd deprecation warnings.

## Change log

```
Added:

* `hoomd.Snapshot.from_gsd_frame` - convert a `gsd.hoomd.Frame` object to `hoomd.Snapshot`.

Deprecated:

* ``hoomd.Snapshot.from_gsd_snapshot`` - use `hoomd.Snapshot.from_gsd_frame`.
```

## Checklist:

- [x] I have reviewed the [**Contributor Guidelines**](https://github.com/glotzerlab/hoomd-blue/blob/trunk-minor/CONTRIBUTING.rst).
- [x] I agree with the terms of the [**HOOMD-blue Contributor Agreement**](https://github.com/glotzerlab/hoomd-blue/blob/trunk-minor/ContributorAgreement.md).
- [x] My name is on the list of contributors (`sphinx-doc/credits.rst`) in the pull request source branch.

### CODE REVIEW

File: hoomd/simulation.py [hunk 1]
```diff
@@ -246,7 +246,12 @@ def create_state_from_snapshot(self,
             # snapshot is hoomd.Snapshot
             self._state = State(self, snapshot, domain_decomposition)
         elif _match_class_path(snapshot, 'gsd.hoomd.Snapshot'):
-            # snapshot is gsd.hoomd.Snapshot
+            # snapshot is gsd.hoomd.Snapshot (gsd 2.x)
+            snapshot = Snapshot.from_gsd_snapshot(snapshot,
+                                                  self._device.communicator)
```
Comment (tommy-waltmann) [inline hoomd/simulation.py]: I suppose the name of this factory method will have to be a little awkward for a while. We could change it to `from_gsd_frame` once hoomd supports only gsd>=2.8.0. Thoughts?

Comment (joaander) [reply hoomd/simulation.py]: Good point. Should we start requiring gsd 2.8.0 in HOOMD-blue v4? Or delay until v5?

For the latter, we can add `from_gsd_frame` now and deprecate `from_gsd_snapshot` for future removal in v5.

Comment (tommy-waltmann) [reply hoomd/simulation.py]: I like adding `from_gsd_frame` and deprecating `from_gsd_snapshot` in v4, removing `from_gsd_snapshot` and requiring `gsd>=2.8.0` in v5. Depending on when a v5 is released, we could require `gsd>=3.0.0` in v5.

**Review by tommy-waltmann — APPROVED** [review]
LGTM

**Pull request merged**